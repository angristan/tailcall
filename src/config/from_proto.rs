#![allow(dead_code)] // TODO check what to do..

use std::collections::BTreeMap;

use convert_case::{Case, Casing};
use prost_reflect::prost_types::{
    DescriptorProto, EnumDescriptorProto, FileDescriptorSet, ServiceDescriptorProto,
};

use crate::config::{Arg, Config, Field, Type};

fn convert_ty(proto_ty: &str) -> String {
    let binding = proto_ty.to_lowercase();
    let proto_ty = binding.strip_prefix("type_").unwrap_or(proto_ty);
    match proto_ty {
        "double" | "float" => "Float",
        "int32" | "int64" | "fixed32" | "fixed64" | "uint32" | "uint64" => "Int",
        "bool" => "Boolean",
        "string" | "bytes" => "String",
        x => x,
    }
    .to_string()
}

fn get_output_ty(output_ty: &str) -> (String, bool) {
    // type, required
    match output_ty {
        "google.protobuf.Empty" => {
            ("String".to_string(), false) // If it's no response is expected, we
                                          // return a nullable string type
        }
        any => (any.to_string(), true), /* Setting it not null by default. There's no way to
                                         * infer this from proto file */
    }
}

fn get_arg(input_ty: &str) -> Option<(String, Arg)> {
    match input_ty {
        "google.protobuf.Empty" | "" => None,
        any => {
            let key = convert_ty(any).to_case(Case::Camel);
            let val = Arg {
                type_of: any.to_string(),
                list: false,
                required: true,
                /* Setting it not null by default. There's no way to infer this
                 * from proto file */
                doc: None,
                modify: None,
                default_value: None,
            };

            Some((key, val))
        }
    }
}

fn append_enums(map: &mut BTreeMap<String, Type>, enums: Vec<EnumDescriptorProto>) {
    for enum_ in enums {
        let ty = Type {
            variants: Some(enum_.value.iter().map(|v| v.name().to_string()).collect()),
            ..Default::default()
        };
        map.insert(enum_.name().to_string(), ty);
    }
}

fn append_msg_type(map: &mut BTreeMap<String, Type>, messages: Vec<DescriptorProto>) {
    if messages.is_empty() {
        return;
    }
    for message in messages {
        let msg_name = message.name().to_string();
        append_enums(map, message.enum_type);
        append_msg_type(map, message.nested_type);

        let mut ty = Type::default();

        for field in message.field {
            let field_name = field.name().to_string();
            let mut cfg_field = Field::default();

            let label = field.label().as_str_name().to_lowercase();
            cfg_field.list = label.contains("repeated");

            if field.r#type.is_some() {
                // for non-primitive types
                let type_of = convert_ty(field.r#type().as_str_name());
                cfg_field.type_of = type_of.to_string();
            } else {
                cfg_field.type_of = convert_ty(field.type_name());
            }

            ty.fields.insert(field_name, cfg_field);
        }
        map.insert(msg_name, ty);
    }
}

fn append_service(
    map: &mut BTreeMap<String, Type>,
    services: Vec<ServiceDescriptorProto>,
    query: String,
) {
    if services.is_empty() {
        return;
    }
    let mut ty = Type::default();

    for service in services {
        for method in service.method {
            let mut cfg_field = Field::default();
            if let Some((k, v)) = get_arg(method.input_type()) {
                cfg_field.args.insert(k, v);
            }

            let (output_ty, required) = get_output_ty(method.output_type());
            cfg_field.type_of = output_ty;
            cfg_field.required = required;
            let name = method.name().to_case(Case::Camel);
            ty.fields.insert(name, cfg_field);
        }
    }
    map.insert(query, ty);
}

pub fn from_proto(
    descriptor_set: FileDescriptorSet,
    query: Option<String>,
) -> anyhow::Result<Config> {
    let mut config = Config::default();
    let mut types = BTreeMap::new();
    let query = query.unwrap_or("Query".to_string());

    for file_descriptor in descriptor_set.file {
        append_enums(&mut types, file_descriptor.enum_type);
        append_msg_type(&mut types, file_descriptor.message_type);
        append_service(&mut types, file_descriptor.service, query.clone());
    }

    config.types = types;

    Ok(config)
}

#[cfg(test)]
mod test { // TODO add proper tests
    use prost_reflect::prost_types::FileDescriptorSet;

    use crate::config::from_proto::from_proto;

    static FOO: &str = r#"
    syntax = "proto3";

        import "google/protobuf/empty.proto";

        package news;

        enum Status {
          PUBLISHED = 0;
          DRAFT = 1;
        }


        message News {
          int32 id = 1;
          string title = 2;
          string body = 3;
          string postImage = 4;
          Status foo = 5;
        }

        service NewsService {
          rpc GetAllNews (google.protobuf.Empty) returns (NewsList) {}
          rpc GetNews (NewsId) returns (News) {}
          rpc GetMultipleNews (MultipleNewsId) returns (NewsList) {}
          rpc DeleteNews (NewsId) returns (google.protobuf.Empty) {}
          rpc EditNews (News) returns (News) {}
          rpc AddNews (News) returns (News) {}
        }

        message NewsId {
          int32 id = 1;
        }

        message MultipleNewsId {
          repeated NewsId ids = 1;
        }

        message NewsList {
          repeated News news = 1;
        }
    "#;

    #[test]
    fn test_from_proto() -> anyhow::Result<()> {
        let mut set = FileDescriptorSet::default();
        let file_desc = protox_parse::parse("news.proto", FOO)?;
        set.file.push(file_desc);

        let result = from_proto(set, None)?;
        println!("{}", result.to_sdl());
        Ok(())
    }
}
