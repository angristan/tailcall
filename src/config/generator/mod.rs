pub mod from_proto;
mod proto_generator;
