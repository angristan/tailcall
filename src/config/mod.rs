pub use config::*;
pub use config_module::*;
pub(crate) use expr::*;
pub(crate) use group_by::*;
pub(crate) use key_values::*;
pub(crate) use link::*;
pub(crate) use reader::*;
pub(crate) use server::*;
pub use source::*;
pub use upstream::*;
mod config;
mod config_module;
mod expr;
mod from_document;
mod group_by;
mod into_document;
mod key_values;
mod link;
mod n_plus_one;
mod reader;
mod server;
mod source;
mod upstream;
