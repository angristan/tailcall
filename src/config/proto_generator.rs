use derive_setters::Setters;

#[derive(Setters)]
pub struct ProtoGeneratorConfig {
    query: String,
    mutation: String,
    is_mutation_fxn: Box<dyn Fn(&str) -> bool>,
}

impl ProtoGeneratorConfig {
    pub fn new(
        query: Option<String>,
        mutation: Option<String>,
        is_mutation_fxn: Box<dyn Fn(&str) -> bool>,
    ) -> Self {
        Self {
            query: query.unwrap_or("Query".to_string()),
            mutation: mutation.unwrap_or("Mutation".to_string()),
            is_mutation_fxn,
        }
    }

    pub fn is_mutation(&self, name: &str) -> bool {
        (self.is_mutation_fxn)(name)
    }
    pub fn get_query(&self) -> &str {
        self.query.as_str()
    }

    pub fn get_mutation(&self) -> &str {
        self.mutation.as_str()
    }
}

impl Default for ProtoGeneratorConfig {
    fn default() -> Self {
        Self {
            query: "Query".to_string(),
            mutation: "Mutation".to_string(),
            is_mutation_fxn: Box::new(|_| false),
        }
    }
}
