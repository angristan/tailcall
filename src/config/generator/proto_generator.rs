use std::collections::BTreeMap;
use std::ops::Deref;
use derive_setters::Setters;
use prost_reflect::prost_types::FileDescriptorSet;
use strum_macros::Display;

use crate::config::{Config, Type};
use crate::config::generator::from_proto::prebuild_config;


#[derive(Default)]
pub struct ConfigWrapper {
    pub config: Config,
    pub types: BTreeMap<String, Vec<String>>
}

impl ConfigWrapper {
    pub fn insert_ty(&mut self, key: String, val: Type, ty: String) {
        if let Some(val) = self.types.get_mut(&ty) {
            val.push(key.clone());
        } else {
            self.types.insert(ty, vec![key.clone()]);
        }

        self.config.types.insert(key, val);
    }
    pub fn get_ty(&self, key: &str) -> Type {
        self.config.types.get(key).cloned().unwrap_or_default()
    }
}

impl Deref for ConfigWrapper {
    type Target = Config;

    fn deref(&self) -> &Self::Target {
        &self.config
    }
}

#[derive(Display, Clone, Copy)]
pub enum DescriptorType {
    Enum,
    Message,
    Method,
}

#[derive(PartialEq, Eq, Copy, Clone)]
pub enum Options {
    // TODO rename
    AppendPkgId,
    FailIfCollide,
    Merge,
}

pub struct ProtoGeneratorFxn {
    is_mutation: Box<dyn Fn(&str) -> bool>,
    format_enum: Box<dyn Fn(Vec<String>) -> BTreeMap<String, String>>,
    format_ty: Box<dyn Fn(Vec<String>) -> BTreeMap<String,String>>,
}

impl ProtoGeneratorFxn {
    pub fn new(
        is_mutation: Box<dyn Fn(&str) -> bool>,
        format_enum: Box<dyn Fn(Vec<String>) ->  BTreeMap<String,String>>,
        format_ty: Box<dyn Fn(Vec<String>) ->  BTreeMap<String,String>>,
    ) -> Self {
        Self {
            is_mutation,
            format_enum,
            format_ty,
        }
    }
}

impl Default for ProtoGeneratorFxn {
    fn default() -> Self {
        let fmt = |x: Vec<String>| {
            let mut map = BTreeMap::new();
            x.into_iter().for_each(|v| { map.insert(v.clone(), v); });
            map
        };
        Self {
            is_mutation: Box::new(|_| false),
            format_enum: Box::new(fmt),
            format_ty: Box::new(fmt),
        }
    }
}

#[derive(Setters)]
pub struct ProtoGeneratorConfig {
    query: String,
    mutation: String,
    generator_fxn: ProtoGeneratorFxn,
    option: Options,
}

impl ProtoGeneratorConfig {
    pub fn new(
        query: Option<String>,
        mutation: Option<String>,
        generator_fxn: ProtoGeneratorFxn,
    ) -> Self {
        Self {
            query: query.unwrap_or_else(|| "Query".to_string()),
            mutation: mutation.unwrap_or_else(|| "Mutation".to_string()),
            generator_fxn,
            option: Options::AppendPkgId,
        }
    }

    pub fn is_mutation(&self, name: &str) -> bool {
        (self.generator_fxn.is_mutation)(name)
    }
    pub fn get_query(&self) -> &str {
        self.query.as_str()
    }

    pub fn get_mutation(&self) -> &str {
        self.mutation.as_str()
    }
}

impl Default for ProtoGeneratorConfig {
    fn default() -> Self {
        Self {
            query: "Query".to_string(),
            mutation: "Mutation".to_string(),
            generator_fxn: ProtoGeneratorFxn::default(),
            option: Options::AppendPkgId,
        }
    }
}
struct ProtoGenerator {
    generator_config: ProtoGeneratorConfig,
}

impl ProtoGenerator {
    pub fn new(generator_config: ProtoGeneratorConfig) -> Self {
        Self { generator_config }
    }

    pub fn generate(&self, descriptor_sets: Vec<FileDescriptorSet>) -> anyhow::Result<Config> {
        let mut config = Config::default();
        let pre_built_wrapper = prebuild_config(descriptor_sets, &self.generator_config, self.generator_config.option)?;
        config.schema = pre_built_wrapper.config.schema;

        match self.generator_config.option {
            Options::AppendPkgId => {

            }
            _ => (),
        }

        Ok(config)
    }
}