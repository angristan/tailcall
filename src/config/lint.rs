use std::collections::{BTreeMap, BTreeSet};

use convert_case::Casing;
use serde::{Deserialize, Serialize};

use crate::config::{AddField, Config, Type};

#[derive(Serialize, Deserialize, Clone, Debug, Default, PartialEq, Eq, schemars::JsonSchema)]
pub struct Lint {
    #[serde(rename = "field")]
    pub field_lint: bool,
    #[serde(rename = "type")]
    pub type_lint: bool,
    #[serde(rename = "enum")]
    pub enum_lint: bool,
    #[serde(rename = "enumValue")]
    pub enum_value_lint: bool,
    #[serde(rename = "autoFix")]
    pub autofix: bool,
}

// Type -> PascalCase
// Field -> camelCase
// Enum -> PascalCase
// Enum Values -> ALL_CAPS

pub fn lint_fix(mut config: Config) -> Config {
    if let Some(lint) = &config.server.lint {
        let mut fixed_types = BTreeMap::new();
        for (k, ty) in config.types.iter() {
            fixed_types.insert(
                match lint.type_lint {
                    true => k.to_case(convert_case::Case::Pascal),
                    false => k.clone(),
                },
                fix_type(lint, ty),
            );
        }
        config.types = fixed_types;
    }
    config
}

// TODO: check variants, fields and added_fields
fn fix_type(lint: &Lint, ty: &Type) -> Type {
    let mut new_ty = Type::default();

    // for enum
    if lint.enum_lint {
        if let Some(variants) = &ty.variants {
            let mut set = BTreeSet::new();
            for variant in variants {
                set.insert(variant.to_case(convert_case::Case::Upper));
            }
            new_ty.variants = Some(set);
        }
    }

    if lint.field_lint {
        for (k, field) in &ty.fields {
            new_ty
                .fields
                .insert(k.to_case(convert_case::Case::Camel), field.clone());
        }
        for field in &ty.added_fields {
            new_ty.added_fields.push(AddField {
                name: field.name.to_case(convert_case::Case::Camel),
                path: field.path.clone(),
            })
        }
    }

    new_ty
}
