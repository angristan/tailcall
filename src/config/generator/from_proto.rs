use std::collections::{BTreeSet, HashMap};

use anyhow::anyhow;
use convert_case::{Case, Casing};
use prost_reflect::prost_types::{
    DescriptorProto, EnumDescriptorProto, FileDescriptorSet, ServiceDescriptorProto,
};

use crate::blueprint::GrpcMethod;
use crate::config::generator::proto_generator::{
    ConfigWrapper, DescriptorType, Options, ProtoGeneratorConfig, DEFAULT_SPECTATOR,
};
use crate::config::{Arg, Field, Grpc, Type};

struct Helper {
    map: HashMap<String, String>,
    package: String,
    options: Options,
}

impl Helper {
    fn with_options(options: Options) -> Self {
        Self { map: Default::default(), package: "".to_string(), options }
    }

    fn get_value(&self, name: &str, ty: DescriptorType) -> String {
        match self.options {
            Options::AppendPkgId => match ty {
                DescriptorType::Enum => {
                    format!("{}{}{}", name, DEFAULT_SPECTATOR, self.package)
                }
                DescriptorType::Message => {
                    format!("{}{}{}", name, DEFAULT_SPECTATOR, self.package)
                }
                DescriptorType::Method => format!(
                    "{}{}{}",
                    name.to_case(Case::Camel),
                    DEFAULT_SPECTATOR,
                    self.package
                ),
            },
            _ => match ty {
                DescriptorType::Enum => name.to_string(),
                DescriptorType::Message => name.to_string(),
                DescriptorType::Method => name.to_case(Case::Camel),
            },
        }
    }
    fn insert(&mut self, name: &str, ty: DescriptorType) -> anyhow::Result<()> {
        if Options::FailIfCollide == self.options {
            if self
                .map
                .keys()
                .any(|v| v.split('.').last().map(|c| c.eq(name)).unwrap_or_default())
            {
                return Err(anyhow!("Duplicate keys found for: {}", name));
            } else {
                self.map.insert(name.to_string(), self.get_value(name, ty));
            }
        } else if Options::Merge == self.options {
            self.map.insert(name.to_string(), self.get_value(name, ty));
        } else {
            self.map.insert(
                format!("{}.{}", self.package, name),
                self.get_value(name, ty),
            );
        }
        Ok(())
    }
    fn get(&self, name: &str) -> Option<String> {
        match self.options {
            Options::AppendPkgId => self.map.get(&format!("{}.{}", self.package, name)).cloned(),
            _ => self.map.get(name).cloned(),
        }
    }
}

fn convert_ty(proto_ty: &str) -> String {
    let binding = proto_ty.to_lowercase();
    let proto_ty = binding.strip_prefix("type_").unwrap_or(proto_ty);
    match proto_ty {
        "double" | "float" => "Float",
        "int32" | "int64" | "fixed32" | "fixed64" | "uint32" | "uint64" => "Int",
        "bool" => "Boolean",
        "string" | "bytes" => "String",
        x => x,
    }
    .to_string()
}

fn get_output_ty(output_ty: &str) -> (String, bool) {
    // type, required
    match output_ty {
        "google.protobuf.Empty" => {
            ("String".to_string(), false) // If it's no response is expected, we
                                          // return a nullable string type
        }
        any => (any.to_string(), true), /* Setting it not null by default. There's no way to
                                         * infer this from proto file */
    }
}

fn get_arg(input_ty: &str, helper: &mut Helper) -> Option<(String, Arg)> {
    match input_ty {
        "google.protobuf.Empty" | "" => None,
        any => {
            let key = convert_ty(any).to_case(Case::Camel);
            let val = Arg {
                type_of: helper.get(any).unwrap_or(any.to_string()),
                list: false,
                required: true,
                /* Setting it not null by default. There's no way to infer this
                 * from proto file */
                doc: None,
                modify: None,
                default_value: None,
            };

            Some((key, val))
        }
    }
}

fn append_enums(
    config_wrapper: &mut ConfigWrapper,
    enums: Vec<EnumDescriptorProto>,
    helper: &mut Helper,
) -> anyhow::Result<()> {
    for enum_ in enums {
        let enum_name = enum_.name();

        let enum_ty = DescriptorType::Enum;
        helper.insert(enum_name, enum_ty)?;

        let mut ty = config_wrapper.get_ty(&helper.get(enum_name).unwrap());

        let mut variants = enum_
            .value
            .iter()
            .map(|v| v.name().to_string())
            .collect::<BTreeSet<String>>();
        if let Some(vars) = ty.variants {
            variants.extend(vars);
        }
        ty.variants = Some(variants);

        config_wrapper.insert_ty(helper.get(enum_name).unwrap(), ty, enum_ty.to_string());
        // it should be
        // safe to call
        // unwrap here
    }
    Ok(())
}

fn append_msg_type(
    config_wrapper: &mut ConfigWrapper,
    messages: Vec<DescriptorProto>,
    helper: &mut Helper,
) -> anyhow::Result<()> {
    if messages.is_empty() {
        return Ok(());
    }
    for message in messages {
        let msg_name = message.name().to_string();

        let msg_ty = DescriptorType::Message;
        helper.insert(&msg_name, msg_ty)?;

        append_enums(config_wrapper, message.enum_type, helper)?;
        append_msg_type(config_wrapper, message.nested_type, helper)?;

        let mut ty = config_wrapper.get_ty(&helper.get(&msg_name).unwrap()); // it should be
                                                                             // safe to call
                                                                             // unwrap here

        for field in message.field {
            let field_name = field.name().to_string();
            let mut cfg_field = Field::default();

            let label = field.label().as_str_name().to_lowercase();
            cfg_field.list = label.contains("repeated");
            cfg_field.required = label.contains("required");

            if field.r#type.is_some() {
                let type_of = convert_ty(field.r#type().as_str_name());
                cfg_field.type_of = type_of.to_string();
            } else {
                // for non-primitive types
                let type_of = convert_ty(field.type_name());
                cfg_field.type_of = helper.get(&type_of).unwrap_or(type_of);
            }

            ty.fields.insert(field_name, cfg_field);
        }

        config_wrapper.insert_ty(helper.get(&msg_name).unwrap(), ty, msg_ty.to_string());
        // it should be
        // safe to call
        // unwrap here
    }
    Ok(())
}

fn generate_ty(
    config_wrapper: &mut ConfigWrapper,
    services: Vec<ServiceDescriptorProto>,
    helper: &mut Helper,
    key: &str,
) -> anyhow::Result<Type> {
    let package = helper.package.clone();
    let mut grpc_method = GrpcMethod { package, service: "".to_string(), name: "".to_string() };
    let mut ty = config_wrapper.get_ty(key);

    for service in services {
        let service_name = service.name().to_string();
        for method in &service.method {
            let method_name = method.name();

            helper.insert(method_name, DescriptorType::Method)?;

            let mut cfg_field = Field::default();
            if let Some((k, v)) = get_arg(method.input_type(), helper) {
                cfg_field.args.insert(k, v);
            }

            let (output_ty, required) = get_output_ty(method.output_type());
            cfg_field.type_of = helper.get(&output_ty).unwrap_or(output_ty.clone());
            cfg_field.required = required;

            grpc_method.service = service_name.clone();
            grpc_method.name = method_name.to_string();

            cfg_field.grpc = Some(Grpc {
                base_url: None,
                body: None,
                group_by: vec![],
                headers: vec![],
                method: grpc_method.to_string(),
            });
            ty.fields
                .insert(helper.get(method_name).unwrap(), cfg_field);
        }
    }
    Ok(ty)
}

fn append_query_service(
    config_wrapper: &mut ConfigWrapper,
    mut services: Vec<ServiceDescriptorProto>,
    gen: &ProtoGeneratorConfig,
    helper: &mut Helper,
) -> anyhow::Result<()> {
    let query = gen.get_query();

    if services.is_empty() {
        return Ok(());
    }

    for service in services.iter_mut() {
        service.method.retain(|v| !gen.is_mutation(v.name()));
    }

    let ty = generate_ty(config_wrapper, services, helper, query)?;

    if ty.ne(&Type::default()) {
        config_wrapper.config.schema.query = Some(query.to_string());
        config_wrapper.insert_ty(query.to_string(), ty, query.to_string());
    }
    Ok(())
}

fn append_mutation_service(
    config_wrapper: &mut ConfigWrapper,
    mut services: Vec<ServiceDescriptorProto>,
    gen: &ProtoGeneratorConfig,
    helper: &mut Helper,
) -> anyhow::Result<()> {
    let mutation = gen.get_mutation();
    if services.is_empty() {
        return Ok(());
    }

    for service in services.iter_mut() {
        service.method.retain(|v| gen.is_mutation(v.name()));
    }

    let ty = generate_ty(config_wrapper, services, helper, mutation)?;
    if ty.ne(&Type::default()) {
        config_wrapper.config.schema.mutation = Some(mutation.to_string());
        config_wrapper.insert_ty(mutation.to_string(), ty, mutation.to_string());
    }
    Ok(())
}

pub fn prebuild_config(
    descriptor_sets: Vec<FileDescriptorSet>,
    gen: &ProtoGeneratorConfig,
    options: Options,
) -> anyhow::Result<ConfigWrapper> {
    let mut config = ConfigWrapper::default();
    let mut helper = Helper::with_options(options);

    for descriptor_set in descriptor_sets {
        for file_descriptor in descriptor_set.file {
            helper.package = file_descriptor.package().to_string();

            append_enums(&mut config, file_descriptor.enum_type, &mut helper)?;
            append_msg_type(&mut config, file_descriptor.message_type, &mut helper)?;
            append_query_service(
                &mut config,
                file_descriptor.service.clone(),
                gen,
                &mut helper,
            )?;
            append_mutation_service(&mut config, file_descriptor.service, gen, &mut helper)?;
        }
    }

    Ok(config)
}

#[cfg(test)]
mod test {

    use std::path::PathBuf;

    use prost_reflect::prost_types::{FileDescriptorProto, FileDescriptorSet};

    use crate::config::generator::from_proto::prebuild_config;
    use crate::config::generator::proto_generator::{
        Options, ProtoGeneratorConfig, ProtoGeneratorFxn,
    };

    fn get_proto_file_descriptor(name: &str) -> anyhow::Result<FileDescriptorProto> {
        let mut proto_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
        proto_path.push("src");
        proto_path.push("grpc");
        proto_path.push("tests");
        proto_path.push("proto");
        proto_path.push(name);
        Ok(protox_parse::parse(
            name,
            std::fs::read_to_string(proto_path)?.as_str(),
        )?)
    }

    fn get_generator_cfg() -> ProtoGeneratorConfig {
        let is_mut = |x: &str| !x.starts_with("Get");
        ProtoGeneratorConfig::new(
            Some("Query".to_string()),
            Some("Mutation".to_string()),
            ProtoGeneratorFxn { is_mutation: Box::new(is_mut), ..Default::default() },
        )
    }

    #[test]
    fn test_from_proto() -> anyhow::Result<()> {
        // news_enum.proto covers:
        // test for mutation
        // test for empty objects
        // test for optional type
        // test for enum
        // test for repeated fields
        // test for a type used as both input and output
        // test for two types having same name in different packages

        let mut set = FileDescriptorSet::default();

        let news = get_proto_file_descriptor("news_enum.proto")?;
        let greetings = get_proto_file_descriptor("greetings.proto")?;
        let greetings_dup_methods = get_proto_file_descriptor("greetings_dup_methods.proto")?;

        set.file.push(news.clone());
        set.file.push(greetings.clone());
        set.file.push(greetings_dup_methods.clone());

        let result =
            prebuild_config(vec![set], &get_generator_cfg(), Options::AppendPkgId)?.to_sdl();

        insta::assert_snapshot!(result);

        // test for 2 different sets
        let mut set = FileDescriptorSet::default();
        let mut set1 = FileDescriptorSet::default();
        let mut set2 = FileDescriptorSet::default();
        set.file.push(news);
        set1.file.push(greetings);
        set2.file.push(greetings_dup_methods);

        let result_sets = prebuild_config(
            vec![set, set1, set2],
            &get_generator_cfg(),
            Options::AppendPkgId,
        )?
        .to_sdl();

        assert_eq!(result, result_sets);

        Ok(())
    }

    #[test]
    fn test_merge() -> anyhow::Result<()> {
        let mut set = FileDescriptorSet::default();

        let news = get_proto_file_descriptor("news_enum.proto")?;
        let greetings = get_proto_file_descriptor("greetings.proto")?;
        let greetings_dup_methods = get_proto_file_descriptor("greetings_dup_methods.proto")?;

        set.file.push(news.clone());
        set.file.push(greetings.clone());
        set.file.push(greetings_dup_methods.clone());

        let result = prebuild_config(vec![set], &get_generator_cfg(), Options::Merge)?.to_sdl();

        insta::assert_snapshot!(result);
        Ok(())
    }

    #[test]
    fn test_fail_if_collide() -> anyhow::Result<()> {
        let mut set = FileDescriptorSet::default();

        let news = get_proto_file_descriptor("news_enum.proto")?;
        let greetings = get_proto_file_descriptor("greetings.proto")?;
        let greetings_dup_methods = get_proto_file_descriptor("greetings_dup_methods.proto")?;

        set.file.push(news.clone());
        set.file.push(greetings.clone());
        set.file.push(greetings_dup_methods.clone());

        let result = prebuild_config(vec![set], &get_generator_cfg(), Options::FailIfCollide);
        assert!(result.is_err());
        Ok(())
    }

    #[test]
    fn test_required_types() -> anyhow::Result<()> {
        // required fields are deprecated in proto3 (https://protobuf.dev/programming-guides/dos-donts/#add-required)
        // this example uses proto2 to test the same.
        // for proto3 it's guaranteed to have a default value (https://protobuf.dev/programming-guides/proto3/#default)
        // and our implementation (https://github.com/tailcallhq/tailcall/pull/1537) supports default values by default.
        // so we do not need to explicitly mark fields as required.

        let mut set = FileDescriptorSet::default();
        let req_proto = get_proto_file_descriptor("required_fields.proto")?;
        set.file.push(req_proto);

        let cfg =
            prebuild_config(vec![set], &get_generator_cfg(), Options::FailIfCollide)?.to_sdl();
        insta::assert_snapshot!(cfg);

        Ok(())
    }
}
