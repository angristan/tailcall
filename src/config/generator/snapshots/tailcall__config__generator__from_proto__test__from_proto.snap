---
source: src/config/generator/from_proto.rs
expression: result
---
schema @server @upstream {
  query: Query
  mutation: Mutation
}

input Author_News {
  email: String
  name: String
}

input HelloRequest_Greetings {
  name: String
}

input HelloRequest_GreetingsDup {
  name: String
}

input MultipleNewsId_News {
  ids: [NewsId_News]
}

input NewsId_News {
  id: Int
}

input News_News {
  author: Author_News
  body: String
  foo: Status_NEWS
  id: Int
  postImage: String
  title: String
}

enum Status_NEWS {
  DRAFT
  PUBLISHED
}

type HelloReply_Greetings {
  message: String
}

type HelloReply_GreetingsDup {
  message: String
}

type Mutation {
  addNews_News(news: News_News!): News_News! @grpc(method: "news.NewsService.AddNews")
  deleteNews_News(newsId: NewsId_News!): String @grpc(method: "news.NewsService.DeleteNews")
  editNews_News(news: News_News!): News_News! @grpc(method: "news.NewsService.EditNews")
  sayHello_Greetings(helloRequest: HelloRequest_Greetings!): HelloReply_Greetings! @grpc(method: "greetings.Greeter.SayHello")
  sayHello_GreetingsDup(helloRequest: HelloRequest_GreetingsDup!): HelloReply_GreetingsDup! @grpc(method: "greetings_dup.Greeter.SayHello")
}

type NewsList_News {
  news: [News_News]
}

type Query {
  getAllNews_News: NewsList_News! @grpc(method: "news.NewsService.GetAllNews")
  getMultipleNews_News(multipleNewsId: MultipleNewsId_News!): NewsList_News! @grpc(method: "news.NewsService.GetMultipleNews")
  getNews_News(newsId: NewsId_News!): News_News! @grpc(method: "news.NewsService.GetNews")
}
